File: algorithms/2_medium/M/03397_maximum_number_of_distinct_elements_after_operations/solution.cpp
class Solution {
public:
    int maxDistinctElements(vector<int>& nums, int k) {

        sort(nums.begin(), nums.end());

        int last_distinct{INT_MIN};
        int count{0};

        for (const int &num: nums) {
            const int max_possible{num + k};

            if (last_distinct < max_possible) {
                last_distinct = max(last_distinct + 1, num - k);
                count += 1;
            }
        }

        return count;
    }
};

File: algorithms/2_medium/M/03397_maximum_number_of_distinct_elements_after_operations/solution.py
INT_MIN = -1 - (2 ** 31)

class Solution:
    def maxDistinctElements(self, nums: List[int], k: int) -> int:

        nums.sort()

        last_distinct = INT_MIN
        count = 0

        for num in nums:
            max_possible = num + k

            if last_distinct < max_possible:
                last_distinct = max(last_distinct + 1, num - k)
                count += 1

        return count
